equivocation.go

```diff
@@ -0,0 +1,110 @@
+package f3
+
+import (
+	"bytes"
+	"slices"
+	"sync"
+
+	"github.com/filecoin-project/go-f3/gpbft"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+// zero value is valid
+type equivocationFilter struct {
+	lk              sync.Mutex
+	currentInstance uint64
+	// sentMessages map unique message slot to its signature
+	sentMessages  map[equivocationKey][]byte
+	activeSenders map[gpbft.ActorID][]peer.ID
+}
+
+type equivocationKey struct {
+	Sender gpbft.ActorID
+	Round  uint64
+	Phase  gpbft.Phase
+}
+
+func (ef *equivocationFilter) formKey(m gpbft.GMessage) equivocationKey {
+	return equivocationKey{
+		Sender: m.Sender,
+		Round:  m.Vote.Round,
+		Phase:  m.Vote.Phase,
+	}
+}
+
+func (ef *equivocationFilter) ProcessBroadcast(peerID peer.ID, m *gpbft.GMessage) bool {
+	ef.lk.Lock()
+	defer ef.lk.Unlock()
+
+	if m.Vote.Instance < ef.currentInstance {
+		// disallow past instances
+		log.Debugw("disallowing broadcast for past message", "sender", m.Sender, "instance",
+			m.Vote.Instance, "currentInstance", ef.currentInstance)
+		return false
+	}
+	// lazy initialization or moved onto new instance
+	if ef.sentMessages == nil || m.Vote.Instance > ef.currentInstance {
+		// bump to next instance
+		ef.currentInstance = m.Vote.Instance
+		ef.sentMessages = make(map[equivocationKey][]byte)
+		ef.activeSenders = make(map[gpbft.ActorID][]peer.ID)
+	}
+
+	key := ef.formKey(*m)
+	sig, ok := ef.sentMessages[key]
+	if ok {
+		log.Debugw("disallowing broadcast due to signature mismatch", "sender", m.Sender,
+			"instance", m.Vote.Instance, "round", m.Vote.Round, "phase", m.Vote.Phase)
+		return bytes.Equal(sig, m.Signature)
+	}
+
+	// save the signature
+	ef.sentMessages[key] = m.Signature
```

Comment:
Thinking out loud: It would be nice to avoid double storing messages somehow since we store messages already in validation cache deep inside gpbft. 

In the meantime, a comment on the filter struct to document the memory complexity would be much appreciated.

Reply:
We only store messages we send, which is very little.